HZN-1331: Consume events from Kafka

malateshs49 opened the pull request:
This PR includes -

- Ability to consume and process generic events from Kafka inside eventd.
- We can now add threads, batchsize and queue size as part of eventd-configuration.xml now for eventsinkconsumer module(configurable).

Jira Link - https://issues.opennms.org/browse/HZN-1331

pbrane commented:
Could you please create an issue in our Jira describing this feature enhancement?  It's the only way we have to track the contributions, create release-notes, etc. etc.

Thanks,
David

malateshs49 commented:
Please find the JIRA created for the new feature done -
https://issues.opennms.org/browse/BSHIFT-176
Please let me know if anything else needed.

j-white commented on features/events/api/src/main/java/org/opennms/netmgt/events/api/Events.java:
```diff
@@ -0,0 +1,66 @@
+import com.google.common.primitives.Bytes;
+
+@XmlRootElement(name = "event")
+public class Events implements Message {
```
Would it be possible to use existing `org.opennms.netmgt.xml.event.Log` and `org.opennms.netmgt.xml.event.Event` objects instead of defining new ones?

malateshs49 replied:
Yes, org.opennms.netmgt.xml.event.Log is a very good option. Rather than converting logs at receiving end i would change these accordingly by updating it.Thanks for the info

malateshs49 replied:
@j-white : Sorry jesse i got carried away.why new Events class was written because org.opennms.netmgt.xml.event.Log or org.opennms.netmgt.xml.event.Event doesn't implement org.opennms.core.ipc.sink.api.Message hence had to write new Events class in order to use MessageConsumer

j-white replied:
Can we just implement those interfaces on the existing objects?

j-white commented on opennms-base-assembly/src/main/filtered/etc/eventd-configuration.xml:
```diff
@@ -1,2 +1,2 @@
-<EventdConfiguration xmlns="http://xmlns.opennms.org/xsd/config/eventd" TCPAddress="127.0.0.1" TCPPort="5817" UDPAddress="127.0.0.1" UDPPort="5817" receivers="5" getNextEventID="SELECT nextval('eventsNxtId')" socketSoTimeoutRequired="yes" socketSoTimeoutPeriod="3000">
+<EventdConfiguration xmlns="http://xmlns.opennms.org/xsd/config/eventd" TCPAddress="127.0.0.1" TCPPort="5817" UDPAddress="127.0.0.1" UDPPort="5817" receivers="5" getNextEventID="SELECT nextval('eventsNxtId')" socketSoTimeoutRequired="yes" socketSoTimeoutPeriod="3000" threads="300">
```
Can we avoid modifying this file? Other sink modules default to (2*number of cores) for the number of threads when none are set.

malateshs49 replied:
Yes thats the main what we have, this field is like configurable like how syslogd-configuration.xml and trapd-configuration.xml has where we can even make threads configurable.

malateshs49 commented:
@j-white : Sure i will take care of the copyright year changes and will update the code and push it back.I just followed previous copyright.

j-white commented on features/events/api/pom.xml:
```diff
@@ -83,6 +83,10 @@
     </dependency>
+    <dependency>
+      <groupId>org.opennms.core.ipc.sink</groupId>
+      <artifactId>org.opennms.core.ipc.sink.xml</artifactId>
```
We should change this to depend on the `org.opennms.core.ipc.sink.api` instead, and add a dependency on `org.opennms.core.ipc.sink.xml` in the `features/events/daemon/pom.xml` where the module is implemented.

j-white commented on features/events/api/src/main/java/org/opennms/netmgt/events/api/EventsWrapper.java:
```diff
@@ -0,0 +1,52 @@
+    private final Log eventLog;
+
+    public EventsWrapper(Log eventLog)
+            throws IllegalArgumentException {
```
We never actually throw a `IllegalArgumentException`.

malateshs49 replied:
our old files had illegalArgument the same came off here.

j-white commented on opennms-config-model/src/main/java/org/opennms/netmgt/config/eventd/EventdConfiguration.java:
```diff
@@ -127,6 +127,36 @@
+     * Number of threads used for consuming/dispatching messages.
+     * Defaults to 2 x the number of available processors.
+     */
+    @XmlAttribute(name = "threads", required=false)
```
Can we prefix these new attribute with `sink-` to make it clear that these settings are used for the Sink API consumer?

malateshs49 replied:
we followed the legacy of how it is handled and described in syslogd-configuration.xml and trapd-configuration.xml for example https://github.com/OpenNMS/opennms/blob/develop/opennms-base-assembly/src/main/filtered/etc/trapd-configuration.xml

j-white replied:
In those cases all of the traps/syslogs are sourced from the Sink API, so there's no ambiguity as to what threads we're configuring. In the case of eventd there are a number of existing threads pools, and various sources of events. Prefixing these attributes will help make whats being configured clearer.

j-white commented on opennms-config-model/src/main/java/org/opennms/netmgt/config/eventd/EventdConfiguration.java:
```diff
@@ -162,6 +192,14 @@ public Integer getUDPPort() {
+    }
+
+    public void setNumThreads(final Integer numThreads) {
+        m_threads = ConfigUtils.assertNotNull(numThreads, "threads");
```
Why do we not allow this field to be null?

malateshs49 replied:
We have maintained same legacy of opennms i.e syslogd it self which is actually kept as not null  for example https://github.com/OpenNMS/opennms/blob/develop/opennms-config-model/src/main/java/org/opennms/netmgt/config/syslogd/Configuration.java .If we need to change then we are ready to update it.

j-white replied:
In the case referenced we perform validation on *all* of the attributes, whereas here we only verify that the `threads` field is not null. The behavior is not consistent.

j-white commented on features/events/daemon/src/main/java/org/opennms/netmgt/eventd/sink/EventsModule.java:
```diff
@@ -0,0 +1,136 @@
+
+			@Override
+			public Log aggregate(Log accumulator, EventsWrapper eventsWrapper) {
+				accumulator = eventsWrapper.getEvents();
```
The  aggregate method should add the given object to the existing accumulator, and create a new one if necessary. Here we are ignoring the existing accumulator and just using the log in the message.

The `EventsWrapper`  probably be changed to wrap a single event, and this message should updated to roll those up into the `Log`.

malateshs49 replied:
this has been addressed.

j-white commented on features/events/daemon/src/main/java/org/opennms/netmgt/eventd/sink/EventsModule.java:
```diff
@@ -0,0 +1,140 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EventsModule extends AbstractXmlSinkModule<EventsWrapper, Log> {
```
Since the `Log` implements the `Message` interface, it doesn't look like there's any need for the `EventsWrapper` wrapper.

That said, I think we'd be better off using an `Event` as the initial message rather than a `Log` here. Otherwise we need to create a `Log` every time we need to send a single event.

This won't change anything for your specific use case, but will make it easier to use when dispatching messages directly with the Sink API.

malateshs49 replied:
Yup i would update to event from log rather than doing to log at sender side its better to reuse and do at receiver under sendnowsync as usual.However we need eventswrapper class because MessageConsumer<S extends Message, T extends Message> uses two Message argument and we cant use Events or Log as first one since both return Events and in aggregate we cant iterate over Events and get single event or we cant hardcode to return first event .Hence EventsWrapper is used.

j-white commented on opennms-base-assembly/src/main/filtered/etc/eventd-configuration.xml:
```diff
@@ -1,2 +1,2 @@
-<EventdConfiguration xmlns="http://xmlns.opennms.org/xsd/config/eventd" TCPAddress="127.0.0.1" TCPPort="5817" UDPAddress="127.0.0.1" UDPPort="5817" receivers="5" getNextEventID="SELECT nextval('eventsNxtId')" socketSoTimeoutRequired="yes" socketSoTimeoutPeriod="3000">
+<EventdConfiguration xmlns="http://xmlns.opennms.org/xsd/config/eventd" TCPAddress="127.0.0.1" TCPPort="5817" UDPAddress="127.0.0.1" UDPPort="5817" receivers="5" getNextEventID="SELECT nextval('eventsNxtId')" socketSoTimeoutRequired="yes" socketSoTimeoutPeriod="3000" sink-threads="300">
```
Can we omit the entry for `sink-threads` here? We have a default value for this attribute.

j-white commented on opennms-config-model/src/main/java/org/opennms/netmgt/config/eventd/EventdConfiguration.java:
```diff
@@ -127,6 +127,36 @@
+     * Number of threads used for consuming/dispatching messages.
+     * Defaults to 2 x the number of available processors.
+     */
+    @XmlAttribute(name = "sink-threads", required=false)
```
Can we prefix **all** the sink related settings we added with the 'sink-' prefix?

j-white commented on features/events/daemon/src/main/java/org/opennms/netmgt/eventd/sink/EventsModule.java:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	@Override
+	public AggregationPolicy<EventsWrapper, Event, Event> getAggregationPolicy() {
```
In the latest update we switched from using the `Log` to the `Event` for *both* types - this was not the intent of my last comment.

In the current incarnation the `Event` class implements the `Message` interface so there is no need for the  `EventsWrapper`. The `AggregationPolicy` also does not perform any aggregation, so it could be replaced with an instance of the `IdentityAggregationPolicy`. The batch size / interface configurations are also no longer required.

I suggest that we define the module as:
```
public class EventsModule extends AbstractXmlSinkModule<Event, Log> 
```
and update the aggregation policy to append the events to the log, creating a new one when necessary.

RangerRick commented on features/events/api/src/main/java/org/opennms/netmgt/xml/event/Event.java:
```diff
@@ -56,14 +56,15 @@
 //@ValidateUsing("event.xsd")
-public class Event implements Serializable {
+public class Event implements Message,Serializable {
         private static final long serialVersionUID = 6997816158234653400L;
```
If you're changing the interfaces it implements, you need to update the `serialVersionUID`.

RangerRick commented on features/events/api/src/main/java/org/opennms/netmgt/xml/event/Log.java:
```diff
@@ -50,7 +52,7 @@
 //@ValidateUsing("event.xsd")
-public class Log implements Serializable {
+public class Log implements Message,Serializable {
     private static final long serialVersionUID = 8526177705077223094L;
```
here too

RangerRick commented on features/events/api/src/main/java/org/opennms/netmgt/events/api/EventsWrapper.java:
```diff
@@ -0,0 +1,51 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2016 The OpenNMS Group, Inc.
```
Make sure you update the copyrights in the files you create/change if possible, as well.

j-white commented on features/events/api/src/main/java/org/opennms/netmgt/events/api/EventsWrapper.java:
```diff
@@ -0,0 +1,51 @@
+import org.opennms.core.ipc.sink.api.Message;
+import org.opennms.netmgt.xml.event.Event;
+
+public class EventsWrapper implements Message {
```
We can remove this class, since it is no longer used.

j-white commented:
Thanks again for the updates. Pushed to `pr/jira/HZN-1331` for a Bamboo run.

j-white commented:
It looks like the `EventdConfigurationTest` is failing with these changes. See https://bamboo.opennms.org/browse/OPENNMS-ONMS2283-UNIT-2

j-white commented:
That test passed, but there are a few other integration tests failing now.

The system tests are also failing, it looks like there is a error is karaf.log that prevents some of the OSGi features from bootstrapping.

malateshs49 commented:
@j-white : I am confused how does this change impact there.Because we built RPM and did certification testing as well.

malateshs49 commented:
@j-white : You can merge changes from this branch to PR-1331 and let me know if bamboo succeeds.
@cgorantla : Thanks for the quick help.Probably we had old develop code from which we had updated and did PR and later on these new things we added to develop which was missed and caused failure and again Thanks for the quick help.

j-white commented on integration-tests/config/src/test/java/org/opennms/netmgt/config/WillItUnmarshalIT.java:
```diff
@@ -220,7 +220,7 @@ private static void addFile(final Source source, final String file, final Class<
         addFile(Source.CONFIG, "enlinkd-configuration.xml", EnlinkdConfiguration.class, false, null);
         addFile(Source.CONFIG, "eventconf.xml", Events.class, true, null);
-        addFile(Source.CONFIG, "eventd-configuration.xml", EventdConfiguration.class, true, null);
+        addFile(Source.CONFIG, "eventd-configuration.xml", EventdConfiguration.class, false, null);
```
We shouldn't need to disable this check here. If the test is failing, we need to rework how the defaults are handled in the JAXB objects.

cgorantla commented:
@malateshs49  We should add integration-test and docs for this feature. You can refer to [these](https://github.com/OpenNMS/opennms/tree/develop/core/ipc/sink/kafka/itests/src/test/java/org/opennms/core/ipc/sink/kafka/itests) tests

cgorantla commented on opennms-config/src/main/java/org/opennms/netmgt/config/EventdConfigManager.java:
```diff
@@ -242,4 +242,48 @@ public String getGetNextEventID() {
+    public int getNumThreads() {
+        getReadLock().lock();
+        try {
+            if (m_config.getNumThreads() <= 0) {
```
I think this condition should be opposite.

RangerRick commented:
I just noticed I had old "requested changes" that never got approved so this was still showing the red check.  @cgorantla you've been managing this PR, right?  I'll leave this to you, I'm just doing some going through of the current queue.  :)

cgorantla commented:
@malateshs49  I have added few more changes in cg/jira/HZN-1331 branch and bamboo is passing.
Please merge them in your branch to reflect in PR.

cgorantla commented:
@j-white  I think most of the issues are addressed now. Can you please review one more time

cgorantla commented:
Closing this PR as we merged code from cg/jira/HZN-1331
